### PR TITLE
Feature - add error states on the login page

### PR DESCRIPTION
**Pull request by rikhilrai**
Again, tried to copy behaviour on this page [google](https://accounts.google.com/signup/v2/webcreateaccount?flowName=GlifWebSignIn&flowEntry=SignUp) as much as possible. 

I like the idea of displaying more than one error on the form

Screenshot 2021-01-04 at 21 56 42

Screenshot 2021-01-04 at 22 03 02

One thing they don't do on Google is hide the error on user typing again

### CODE REVIEW

File: app/src/pages/loginPage/loginPage.tsx [hunk 1]
```diff
@@ -13,20 +12,53 @@ export interface ILoginPageProps {
...
+    // Regex copied from https://emailregex.com/
+    const emailRegex = new RegExp(/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/);
+    if (email && !emailRegex.test(String(email).toLowerCase())) {
+      setEmailError('Please enter a valid email address');
+    }
+    if (!password) {
+      setPasswordError('Enter a password');
+    }
+    if (password && password.length < 8) {
+      setPasswordError('Use 8 characters or more for your password');
```
Comment (rikhilrai) [inline app/src/pages/loginPage/loginPage.tsx]: Just noticed this expands the password box

File: app/src/pages/loginPage/loginPage.tsx [hunk 2]
```diff
@@ -13,20 +12,53 @@ export interface ILoginPageProps {
...
-      // eslint-disable-next-line no-console
-      console.log('email', email, 'password', password);
-    }).catch((error: KibaException): void => {
-      console.error('error', error);
-    });
+    if (!email) {
+      setEmailError('Enter email address');
+    }
+    // Regex copied from https://emailregex.com/
+    const emailRegex = new RegExp(/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/);
```
Comment (rikhilrai) [inline app/src/pages/loginPage/loginPage.tsx]: not too keen on email regex, but i think overall it's better than not having it. what are your thoughts?

File: app/src/pages/loginPage/loginPage.tsx [hunk 3]
```diff
@@ -13,20 +12,53 @@ export interface ILoginPageProps {
...
+    if (email && !emailRegex.test(String(email).toLowerCase())) {
+      setEmailError('Please enter a valid email address');
```
Comment (rikhilrai) [inline app/src/pages/loginPage/loginPage.tsx]: i don't think this condition is hit. some other thing pops up before any of this code is run in the case of an invalid email address. are you aware of this?

File: app/src/pages/registrationPage/registrationPage.tsx [hunk 4]
```diff
@@ -19,14 +18,15 @@ export const RegistrationPage = (props: IRegistrationPageProps): React.ReactElem
   const [retypedPassword, setRetypedPassword] = React.useState<string | null>(null);
   const [isLoading, setIsLoading] = React.useState<boolean>(false);
 
-  const onRegisterClicked = (): void => {
+  const onRegisterClicked = async (): Promise<void> => {
```
Comment (rikhilrai) [inline app/src/pages/registrationPage/registrationPage.tsx:21]: this small change fixes the loading overlay on the registration page

File: app/src/pages/loginPage/loginPage.tsx (hunk 1 above)
Comment (rikhilrai) [reply app/src/pages/loginPage/loginPage.tsx]: either we can increase the width of the input boxes (accounts for longer email addresses as well) OR we can make this message shorter

Comment (rikhilrai) [reply app/src/pages/loginPage/loginPage.tsx]: ok, i updated the message, we can also increase the width of the input boxes

File: app/src/pages/loginPage/loginPage.tsx (hunk 2 above)
Comment (krishan711) [reply app/src/pages/loginPage/loginPage.tsx]: i think would be nice to move it to a util function or something so its not just floating around here (and we can use it in future projects)

File: app/src/pages/loginPage/loginPage.tsx [hunk 5]
```diff
@@ -39,9 +76,9 @@ export const LoginPage = (props: ILoginPageProps): React.ReactElement => {
             <Stack isFullHeight={true} isFullWidth={true} shouldAddGutters={true} direction={Direction.Vertical} childAlignment={Alignment.Center}>
               <Text variant='header3'>Welcome Back :)</Text>
               <Spacing variant={PaddingSize.Wide} />
-              <SingleLineInput placeholderText='Email Address' inputType={InputType.Email} value={email} onValueChanged={setEmail}/>
+              <SingleLineInput inputWrapperVariant={emailError ? 'error' : 'success'} messageText={emailError || undefined} placeholderText='Email Address' inputType={InputType.Email} value={email} onValueChanged={onEmailTyped}/>
               <Spacing variant={PaddingSize.Wide} />
-              <SingleLineInput placeholderText='Password' inputType={InputType.Password} value={password} onValueChanged={setPassword}/>
+              <SingleLineInput inputWrapperVariant={passwordError ? 'error' : 'success'} messageText={passwordError || undefined} placeholderText='Password' inputType={InputType.Password} value={password} onValueChanged={onPasswordTyped}/>
```
Comment (krishan711) [inline app/src/pages/loginPage/loginPage.tsx]: would be good to split these jsx over multiple lines once they have more than one prop. Also I think the `inputWrapperVariant` might need a case where there are no errors or succes? or is the message never defined in that case?

File: app/src/pages/registrationPage/registrationPage.tsx (hunk 4 above)
Comment (krishan711) [reply app/src/pages/registrationPage/registrationPage.tsx:21]: nice

File: app/src/pages/loginPage/loginPage.tsx (hunk 5 above)
Comment (rikhilrai) [reply app/src/pages/loginPage/loginPage.tsx]: i just changed it to be undefined in the case of lack of error. success doesn't really feel right here

File: app/src/util/stringUtil.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+
+export const isEmailValid = (email: string): boolean => {
```
Comment (rikhilrai) [inline app/src/util/stringUtil.ts:2]: not sure if `stringUtil` is the right place for this 😂   Didn't create an `emailUtil` cos not sure what else would be put in there. maybe we can move this to an emailUtil when we have other email things ?

File: app/src/pages/registrationPage/registrationPage.tsx [hunk 7]
```diff
@@ -19,14 +18,15 @@ export const RegistrationPage = (props: IRegistrationPageProps): React.ReactElem
...
     setIsLoading(true);
-    asyncSleep(1000).then((): void => {
+    try {
+      asyncSleep(1000);
       // eslint-disable-next-line no-console
       console.log('email', email, 'password', password);
-    }).catch((error: KibaException): void => {
+    } catch (error) {
```
Comment (rikhilrai) [inline app/src/pages/registrationPage/registrationPage.tsx:27]: not able to type this anymore.

**Review by krishan711 — APPROVED** [review]

**Pull request merged**